File: Source/Extensions/Tessa.Extensions.Default.Client/Workflow/KrProcess/CommandInterpreter/CreateCardViaDocTypeCommandHandler.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Tessa.Cards;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess.ClientCommandInterpreter;
using Tessa.Platform;
using Tessa.Platform.Storage;
using Tessa.UI;

namespace Tessa.Extensions.Default.Client.Workflow.KrProcess.CommandInterpreter
{
    /// <summary>
    /// Обработчик клиентской команды <see cref="DefaultCommandTypes.CreateCardViaDocType"/>.
    /// </summary>
    public sealed class CreateCardViaDocTypeCommandHandler : ClientCommandHandlerBase
    {
        #region Fields

        private readonly IUIHost uiHost;

        #endregion

        #region Constructor

        public CreateCardViaDocTypeCommandHandler(
            IUIHost uiHost)
        {
            this.uiHost = uiHost;
        }

        #endregion

        #region Base Overrides

        /// <inheritdoc/>
        public override async Task Handle(
            IClientCommandHandlerContext context)
        {
            var command = context.Command;
            if (command.Parameters.TryGetValue(KrConstants.Keys.TypeID, out var typeIDObj)
                && typeIDObj is Guid typeID)
            {
                var info = new Dictionary<string, object>
                {
                    [CardHelper.NewCardBilletKey] = command.Parameters.TryGet<byte[]>(KrConstants.Keys.NewCard),
                    [CardHelper.NewCardBilletSignatureKey] = command.Parameters.TryGet<byte[]>(KrConstants.Keys.NewCardSignature),
                };
                
                if (command.Parameters.TryGetValue(KrConstants.Keys.DocTypeID, out var doctypeIDObj)
                    && command.Parameters.TryGetValue(KrConstants.Keys.DocTypeTitle, out var doctypeTitleObj)
                    && doctypeIDObj is Guid doctypeID
                    && doctypeTitleObj is string doctypeTitle)
                {
                    info[KrConstants.Keys.DocTypeID] = doctypeID;
                    info[KrConstants.Keys.DocTypeTitle] = doctypeTitle;
                }
                
                await DispatcherHelper.InvokeInUIAsync(async () =>
                {
                    using ISplash splash = TessaSplash.Create(TessaSplashMessage.CreatingCard);
                    await this.uiHost.CreateCardAsync(
                        typeID,
                        options: new CreateCardOptions
                        {
                            Splash = splash,
                            Info = info,
                        },
                        cancellationToken: context.CancellationToken);
                });
            }
        }

        #endregion
    }
}

File: WebClient SDK/src/default/workflow/krProcess/commandInterpreter/createCardViaDocTypeCommandHandler.ts
import { ClientCommandHandlerBase } from 'tessa/workflow/krProcess/clientCommandInterpreter';
import { IClientCommandHandlerContext } from 'tessa/workflow/krProcess';
import { getTypedFieldValue, createTypedField, DotNetType } from 'tessa/platform';
import { createCard } from 'tessa/ui/uiHost';

export class CreateCardViaDocTypeCommandHandler extends ClientCommandHandlerBase {
  public async handle(context: IClientCommandHandlerContext) {
    const command = context.command;
    const typeId = getTypedFieldValue(command.parameters['TypeID']);

    if (typeId) {
      const info = {
        ['.NewBilletCard']: command.parameters['NewCard'],
        ['.NewBilletCardSignature']: command.parameters['NewCardSignature']
      };

      const docTypeId = getTypedFieldValue(command.parameters['docTypeID']);
      const docTypeTitle = getTypedFieldValue(command.parameters['docTypeTitle']);

      if (docTypeId && docTypeTitle) {
        info['docTypeID'] = createTypedField(docTypeId, DotNetType.Guid);
        info['docTypeTitle'] = createTypedField(docTypeTitle, DotNetType.String);
      }
      await createCard({
        cardTypeId: typeId,
        info
      });
    }
  }
}
